components/tikv_util/src/lru.rs

```diff
@@ -0,0 +1,415 @@
+    suture(record.prev.as_mut(), record.next.as_mut())
+}
+
+impl<K> Trace<K> {
+    fn new(sample_mask: usize) -> Trace<K> {
+        unsafe {
+            let mut head = Box::new(Record {
+                prev: NonNull::new_unchecked(1usize as _),
+                next: NonNull::new_unchecked(1usize as _),
+                key: MaybeUninit::uninit(),
+            });
+            let mut tail = Box::new(Record {
+                prev: NonNull::new_unchecked(1usize as _),
+                next: NonNull::new_unchecked(1usize as _),
+                key: MaybeUninit::uninit(),
+            });
+            suture(&mut head, &mut tail);
+
+            Trace {
+                head,
+                tail,
+                sample_mask,
+                tick: 0,
+            }
+        }
+    }
+
+    #[inline]
+    fn maybe_promote(&mut self, record: NonNull<Record<K>>) {
+        self.tick += 1;
+        if self.tick & self.sample_mask == 0 {
+            self.promote(record);
+        }
+    }
+
+    fn promote(&mut self, mut record: NonNull<Record<K>>) {
+        unsafe {
+            cut_out(record.as_mut());
+            suture(record.as_mut(), &mut self.head.next.as_mut());
+            suture(&mut self.head, record.as_mut());
+        }
+    }
+
+    fn delete(&mut self, mut record: NonNull<Record<K>>) {
+        unsafe {
+            cut_out(record.as_mut());
+
+            ptr::drop_in_place(Box::from_raw(record.as_ptr()).key.as_mut_ptr());
+        }
+    }
+
+    fn create(&mut self, key: K) -> NonNull<Record<K>> {
+        let record = Box::leak(Box::new(Record {
+            prev: unsafe { NonNull::new_unchecked(&mut *self.head) },
+            next: self.head.next,
+            key: MaybeUninit::new(key),
+        }))
+        .into();
+        unsafe {
+            self.head.next.as_mut().prev = record;
+            self.head.next = record;
+        }
+        record
+    }
+
+    fn reuse_tail(&mut self, key: K) -> (K, NonNull<Record<K>>) {
+        unsafe {
+            let mut record = self.tail.prev;
+            cut_out(record.as_mut());
+            suture(record.as_mut(), self.head.next.as_mut());
+            suture(&mut self.head, record.as_mut());
+
+            let old_key = record.as_mut().key.as_ptr().read();
+            record.as_mut().key = MaybeUninit::new(key);
+            (old_key, record)
+        }
+    }
+
+    fn clear(&mut self) {
+        let mut cur = self.head.next;
+        unsafe {
+            while cur.as_ptr() != &mut *self.tail {
+                let tmp = cur.as_mut().next;
+                ptr::drop_in_place(Box::from_raw(cur.as_ptr()).key.as_mut_ptr());
+                cur = tmp;
+            }
+            suture(&mut self.head, &mut self.tail);
+        }
+    }
+
+    fn remove_tail(&mut self) -> K {
+        unsafe {
+            let mut record = self.tail.prev;
+            cut_out(record.as_mut());
+
+            let r = Box::from_raw(record.as_ptr());
+            r.key.as_ptr().read()
```

Comment:
Why not use `ptr::drop_in_place` here?

Reply:
Because `key` is moved and returned, then freed outside the function.